### PR TITLE
Making config map configuration dynamic, adding fswatcher for config file

### PR DESCRIPTION
**Pull request by tsorya**
**What this PR does / why we need it**:
Make configmap optional and creating fs watch to it.
Configmap must be created from file another way it will add file per volume
kubectl -n nmstate create configmap nmstate-config --from-file=nmstate-config.yaml 
 
**Special notes for your reviewer**:
Tests needs to be added if it will be relevant
**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/manager/main.go [hunk 1]
```diff
@@ -93,7 +95,11 @@ func main() {
 		log.Error(err, "Failed to get watch namespace")
 		os.Exit(1)
 	}
-
+	err = configurations.ConfigWatcher()
```
Comment (phoracek) [inline cmd/manager/main.go]: Please add empty space above and below this newly added code.

File: cmd/manager/main.go [hunk 2]
```diff
@@ -16,7 +16,9 @@ import (
 	"k8s.io/client-go/rest"
 
 	"github.com/nmstate/kubernetes-nmstate/pkg/apis"
+	"github.com/nmstate/kubernetes-nmstate/pkg/configurations"
 	"github.com/nmstate/kubernetes-nmstate/pkg/controller"
+
```
Comment (phoracek) [inline cmd/manager/main.go]: Please drop the empty line. These are all part of this package.

File: deploy/operator.yaml [hunk 3]
```diff
@@ -50,30 +50,34 @@ spec:
...
-              valueFrom:
-                configMapKeyRef:
-                  name: nmstate-config
-                  key: interfaces_filter
             - name: ENABLE_PROFILER
               value: "False"
             - name: PROFILER_PORT
               value: "6060"
+            - name: CONFIG_NAME
+              value: "nmstate-config.yaml"
```
Comment (phoracek) [inline deploy/operator.yaml]: nmstate.yaml is enough since it is already under config/. (just a nit)

File: deploy/operator.yaml [hunk 4]
```diff
@@ -50,30 +50,34 @@ spec:
...
           volumeMounts:
           - name: dbus-socket
             mountPath: /run/dbus/system_bus_socket
+          - name: config-volume
+            mountPath: /etc/config
           securityContext:
             privileged: true
+          ports:
```
Comment (phoracek) [inline deploy/operator.yaml]: How is this related to the change?

File: deploy/operator.yaml [hunk 5]
```diff
@@ -131,13 +135,15 @@ spec:
             - name: NODE_NETWORK_STATE_REFRESH_INTERVAL
               valueFrom:
                 configMapKeyRef:
+                  optional: true
```
Comment (phoracek) [inline deploy/operator.yaml]: You need to mount CM as a volume here too (this is worker DS, the one you changed is the master)

File: pkg/configurations/configurations.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+const defaultNodeNetworkStateRefresh = "5"
+
+var (
+	configPath = defaultConfigPath
+	configName = defaultConfigName
+)
+
+// Take os environment values if given else use defaults
+func init() {
+	configPathTemp, isSet := os.LookupEnv("CONFIG_PATH")
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: You put the default for CONFIG_NAME into the manifest, please be consistent and do it with this one too. I think we can drop the defaults from here and just make it required. Require only one parameter - the path to the config, you can get the directory out of it here.

File: pkg/configurations/configurations.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+	}
+	configNameTemp, isSet := os.LookupEnv("CONFIG_NAME")
+	if isSet {
+		configName = configNameTemp
+	}
+}
+
+// Struct to match config file
+type Config struct {
+	NodeNetworkRefreshInterval string `mapstructure:"node_network_state_refresh_interval"`
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: You suffixed the config with .yaml, but here you don't treat it as a yaml. Please let's start with writting docs with examples how to set the config and figure out there what would be the simplest option for users (edit the parameter with a single kubectl command).

File: pkg/helper/client.go [hunk 8]
```diff
@@ -38,16 +38,14 @@ const vlanFilteringCommand = "vlan-filtering"
...
 var (
 	interfacesFilterGlob glob.Glob
 )
 
-func init() {
-	interfacesFilter, isSet := os.LookupEnv("INTERFACES_FILTER")
-	if !isSet {
-		panic("INTERFACES_FILTER is mandatory")
-	}
+func InitClient(interfacesFilter string) {
```
Comment (phoracek) [inline pkg/helper/client.go]: Let's be more specific here and call it SetInterfacesFilter? That's all it does.

File: pkg/helper/config_controller_helper.go [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package helper
```
Comment (phoracek) [inline pkg/helper/config_controller_helper.go]: Why do we still keep config_controller? Did not we obsolete it?

File: pkg/controller/nodenetworkstate/nodenetworkstate_controller.go [hunk 10]
```diff
@@ -118,6 +114,7 @@ func (r *ReconcileNodeNetworkState) Reconcile(request reconcile.Request) (reconc
 			return err
 		}
 		return nil
+
```
Comment (phoracek) [inline pkg/controller/nodenetworkstate/nodenetworkstate_controller.go]: Unrelated change, please drop it

File: pkg/controller/nodenetworkstate/nodenetworkstate_controller.go [hunk 11]
```diff
@@ -29,11 +28,8 @@ var (
 	nodenetworkstateRefresh time.Duration
 )
 
-func init() {
-	refreshTime, isSet := os.LookupEnv("NODE_NETWORK_STATE_REFRESH_INTERVAL")
-	if !isSet {
-		panic("NODE_NETWORK_STATE_REFRESH_INTERVAL is mandatory")
-	}
+func InitNodeNetworkController(refreshTime string) {
```
Comment (phoracek) [inline pkg/controller/nodenetworkstate/nodenetworkstate_controller.go]: Let's call it specifically SetNodeNetworkStateRefreshValue.

File: pkg/configurations/configurations.go (hunk 7 above)
Comment (tsorya) [reply pkg/configurations/configurations.go]: Mapstructure will treat is yaml. Viper will handle it. This is the way they describe using there code.

File: pkg/helper/config_controller_helper.go (hunk 9 above)
Comment (tsorya) [reply pkg/helper/config_controller_helper.go]: We did. Missed it

File: pkg/configurations/configurations.go (hunk 7 above)
Comment (tsorya) [reply pkg/configurations/configurations.go]: Same way we can have json or toml . Just took yaml as default.

**Comment by phoracek** [comment]
@tsorya one other comment. Please make sure to change the value of our global attributes in a sync manner. Wrap it with a Mutex or something.

**Comment by tsorya** [comment]
@phoracek i don't think we need mutex to change simple variable. What do i miss?

**Comment by phoracek** [comment]
I don't think that assignment of a string variable is atomic. If you could reference me a source that says it is, I would be glad to accept it without mutex. But again, it sounds unsafe.

File: pkg/helper/client.go [hunk 12]
```diff
@@ -40,15 +41,15 @@ const defaultGwProbeTimeout = 120 * time.Second
 const apiServerProbeTimeout = 120 * time.Second
 
 var (
+	setMutex             = &sync.Mutex{}
```
Comment (phoracek) [inline pkg/helper/client.go]: Can we move out variables to the configurations/ package? Keep it a global, but private. Expose a structure with Synchronous getters and setters to set and read them. Does it sound good? Put all the configuration on a single space instead of sharing it between modules and put it behind a nice abstraction.

File: pkg/configurations/configurations.go [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+	c.updateConfig(*v)
+	WatchConfig(v, configPath)
+	return nil
+}
+
+//Adapted from viper WatchConfig to match nmstate configmap watch needs
+//The main changes is that there is no need to
+//preexist config file before starting the watch
+//and it will not exit on file deletion
+func WatchConfig(v *viper.Viper, fullFilePath string) {
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: This function can be private

File: pkg/configurations/configurations.go [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+	if err := v.ReadInConfig(); err != nil { // Find and read the config file
+		log.Info("Not able to read configuration, will update default values")
+	}
+
+	c.updateConfig(*v)
+	WatchConfig(v, configPath)
+	return nil
+}
+
+//Adapted from viper WatchConfig to match nmstate configmap watch needs
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: Please add space after `//` to all lines

File: pkg/configurations/configurations.go [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+	if err == nil {
+		return true
+	}
+	if os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+func ConfigWatcher() error {
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: Please add a few lines describing what it does, what it watches, what it changes.

File: pkg/configurations/configurations.go [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: please drop the empty line at the beginning of the function.

File: pkg/configurations/configurations.go [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+func init() {
+	configPathTemp, isSet := os.LookupEnv("CONFIG_PATH")
+	if !isSet {
+		panic("CONFIG_PATH is mandatory")
+	}
+	configPath = configPathTemp
+}
+
+// Struct to match config file
+type Config struct {
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: This can be private. Please add getters (Public) and setters (private) to interface with users and with the watcher. It should be a little cleaner.

File: pkg/configurations/configurations.go [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+const defaultInterfacesFilter = "veth*"
+const defaultNodeNetworkStateRefresh = 5
+
+var (
+	log        = logf.Log.WithName("configurations")
+	configPath string
+)
+
+// Takes os environment values
+func init() {
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: We can get rid of this init (I know, we use them on other places, but it is not a good practice since it randomly panics during imports).

This module should have two structs `Config`, that we already have and `ConfigWatcher`. In main, we should call `NewConfigWatcher`, initializing the structer, that would check the config path. Then we call configWatcher.Start.

Let's try to be clean and safe when dealing with singletons.

File: pkg/configurations/configurations.go [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+	log.Info("Updating configuration")
+	err := v.Unmarshal(&c)
+	log.Info(fmt.Sprintf("New configuration is %+v", c))
+	if err != nil {
+		log.Error(err, "Failed to unmarshal config file")
+		return
+	}
+}
+
+func exists(path string) bool {
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: `exists` is often used in go as a variable. Let's be explicit and call it fileExists

File: pkg/configurations/configurations.go [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+	}
+
+	v.SetConfigFile(configPath)
+	v.SetTypeByDefaultValue(true)
+
+	if err := v.ReadInConfig(); err != nil { // Find and read the config file
+		log.Info("Not able to read configuration, will update default values")
+	}
+
+	c.updateConfig(*v)
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: Why do we need to updateConfig here? Won't it be done by the WatchConfig? This seems redundant

File: pkg/configurations/configurations.go [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+	initWG.Add(1)
+	go func() {
+		newWatcher, err := fsnotify.NewWatcher()
+		if err != nil {
+			log.Error(err, "Failed to start fsnotify watcher")
+			return
+		}
+		defer newWatcher.Close()
+
+		configFile := filepath.Clean(fullFilePath)
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: This seems like initialization that can be done before we enter the goroutine

File: pkg/configurations/configurations.go [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+					eventsWG.Done()
+					return
+				}
+			}
+		}()
+		newWatcher.Add(configDir)
+		initWG.Done()   // done initalizing the watch in this go routine, so the parent routine can move on...
+		eventsWG.Wait() // now, wait for event loop to end in this go-routine...
+	}()
+	initWG.Wait() // make sure that the go routine above fully ended before returning
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: Why do we need the goroutine above? You just execute it and wait for it to end, can't it be dropped?

File: pkg/configurations/configurations.go (hunk 20 above)
Comment (tsorya) [reply pkg/configurations/configurations.go]: It is the case where file already exists. Watch config will run updateConfig only on change

File: pkg/configurations/configurations.go (hunk 22 above)
Comment (tsorya) [reply pkg/configurations/configurations.go]: nope, the above goroutine is added cause of fsnotify watcher. We need to configure it and be able to run defer on if goroutine is finished. I can do it without closing fsnotify watcher but  it looks like it is not a best practice

File: pkg/configurations/configurations.go (hunk 21 above)
Comment (tsorya) [reply pkg/configurations/configurations.go]: why does it matter?

Comment (phoracek) [reply pkg/configurations/configurations.go]: Would make it easier to read. Goroutines add complexity, less they contain, the less I can think about what is happening in parallel. `go` tells me, there will be something running in parallel, I ask "what can go wrong" "what could it confict with"...

File: pkg/configurations/configurations.go (hunk 22 above)
Comment (phoracek) [reply pkg/configurations/configurations.go]: I'm sorry, I don't see it.  If we keep everything, but drop `go func() {` on line 115, what does it change?

File: pkg/helper/client.go [hunk 23]
```diff
@@ -96,7 +85,7 @@ func UpdateCurrentState(client client.Client, nodeNetworkState *nmstatev1alpha1.
 		return fmt.Errorf("error running nmstatectl show: %v", err)
 	}
 	observedState := nmstatev1alpha1.State{Raw: []byte(observedStateRaw)}
-
+	interfacesFilterGlob := glob.MustCompile(configurations.GetCurrentConfig().InterfaceFilter)
```
Comment (phoracek) [inline pkg/helper/client.go]: We can compile it once in before we call the config setter.

File: pkg/configurations/configurations.go (hunk 22 above)
Comment (phoracek) [reply pkg/configurations/configurations.go]: You would be able to drop the initWG too. Then you can keep the defer, it would have exactly the same scope as it has now.

Comment (tsorya) [reply pkg/configurations/configurations.go]: As I checked this is the way to work with fsnotify. U open routine for ents loop and wait for it.

File: pkg/helper/client.go (hunk 23 above)
Comment (tsorya) [reply pkg/helper/client.go]: U want me to add it to configurations? I think it will be a little strange but it's ok

File: pkg/configurations/configurations.go (hunk 22 above)
Comment (phoracek) [reply pkg/configurations/configurations.go]: Example from fsnotify docs: https://godoc.org/github.com/fsnotify/fsnotify#example-NewWatcher

You indeed run a goroutine, I have nothing against the inner one (line 129). The outer one is the issue.

File: pkg/helper/client.go (hunk 23 above)
Comment (phoracek) [reply pkg/helper/client.go]: Meh, no need to.

Comment (tsorya) [reply pkg/helper/client.go]: No it's ok. This function is called A lot. Not sure what is the cost of creating glob but sounds reasonable

File: pkg/configurations/configurations.go (hunk 22 above)
Comment (tsorya) [reply pkg/configurations/configurations.go]: But I need to run the watcher and continue on main thread without closing it. So I create go goroutine and wait in it for inner one.
I think I am missing something. But I don't see how it will work without it.

File: pkg/helper/client.go (hunk 23 above)
Comment (tsorya) [reply pkg/helper/client.go]: So let's decide :)

Comment (phoracek) [reply pkg/helper/client.go]: What makes more sense for you. I don't mind neither.

Comment (tsorya) [reply pkg/helper/client.go]: The question is the cost of glob creation

File: pkg/configurations/configurations.go (hunk 22 above)
Comment (phoracek) [reply pkg/configurations/configurations.go]: You do initWG.Wait(), so you wait until the outer one is done. So it runs the goroutine and then it  blocks forever. You can drop it and wait for eventsWG, would do the same thing. Or maybe I need more sleep!

File: pkg/helper/client.go (hunk 23 above)
Comment (phoracek) [reply pkg/helper/client.go]: Huuuumph, let's do it in configurations then. So there is no regressions (now we call it only once, let's keep that)

**Comment by phoracek** [comment]
@tsorya I managed to set a wrong secret for the prow webhook. We should fix it tomorrow.

**Comment by qinqon** [comment]
@tsorya @phoracek can we use the client-go Watch on configMap so we get notified when they change ? 
https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/core/v1/configmap.go#L97

**Comment by tsorya** [comment]
@qinqon but then it will create controller no? I was told not to create controller.
Actually my first version was with controller

**Comment by phoracek** [comment]
@qinqon I'm against having another controller just for the configmap. It would watch all the CMs in the cluster and require us to run the even watcher. With the current approach, we let Kubernetes update the config actively, we just employ an fsWatcher which is way cheaper and simpler than controller mechanism.

File: pkg/configurations/configurations.go [hunk 24]
```diff
@@ -0,0 +1,159 @@
...
+	realConfigFile, _ := filepath.EvalSymlinks(c.configPath)
+	newWatcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Error(err, "Failed to start fsnotify watcher")
+		return
+	}
+	newWatcher.Add(configDir)
+
+	go func() {
+		defer newWatcher.Close()
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: This should be set right after  the error checking of NewWatcher.

File: pkg/configurations/configurations.go [hunk 25]
```diff
@@ -0,0 +1,159 @@
...
+
+// Adapted from viper WatchConfig to match nmstate configmap watch needs
+// The main changes is that there is no need to
+// preexist config file before starting the watch
+// and it will not exit on file deletion
+func (c *configWatcher) watchConfigFile() {
+	configFile := filepath.Clean(c.configPath)
+	configDir, _ := filepath.Split(configFile)
+	realConfigFile, _ := filepath.EvalSymlinks(c.configPath)
+	newWatcher, err := fsnotify.NewWatcher()
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: Please add an empty line above, to split path handling and initialization of the watcher.

File: pkg/configurations/configurations.go [hunk 26]
```diff
@@ -0,0 +1,159 @@
...
+	if err != nil {
+		log.Error(err, "Failed to start fsnotify watcher")
+		return
+	}
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: Please add an empty line below error handling.

File: pkg/configurations/configurations.go [hunk 27]
```diff
@@ -0,0 +1,159 @@
...
+					realConfigFile = currentConfigFile
+					err := c.v.ReadInConfig()
+					if err != nil {
+						log.Error(err, "error reading config file.\n")
+					} else {
+						setConfig(*c.v)
+					}
+				}
+			case err, ok := <-newWatcher.Errors:
+				if ok { // 'Errors' channel is not closed
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: The default flow should be in the main body and the exceptional one in `if`. Then a reader can do less hops to conditions if they are just following the positive flow. So in this case:

```
if !ok {
    return
}
log.Error(...)

File: pkg/configurations/configurations.go [hunk 28]
```diff
@@ -0,0 +1,159 @@
...
+				// 1 - if the config file was modified or created
+				// 2 - if the real path to the config file changed (eg: k8s ConfigMap replacement)
+				const writeOrCreateMask = fsnotify.Write | fsnotify.Create
+				if (filepath.Clean(event.Name) == configFile &&
+					event.Op&writeOrCreateMask != 0) ||
+					(currentConfigFile != "" && currentConfigFile != realConfigFile) {
+					realConfigFile = currentConfigFile
+					err := c.v.ReadInConfig()
+					if err != nil {
+						log.Error(err, "error reading config file.\n")
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: If err: Log the error and `continue`. Then we can drop the else branch and just handle the positive flow in the main body

File: pkg/configurations/configurations.go [hunk 29]
```diff
@@ -0,0 +1,159 @@
...
+			case event, ok := <-newWatcher.Events:
+				if !ok { // 'Events' channel is closed
+					return
+				}
+				currentConfigFile, _ := filepath.EvalSymlinks(c.configPath)
+				// we only care about the config file with the following cases:
+				// 1 - if the config file was modified or created
+				// 2 - if the real path to the config file changed (eg: k8s ConfigMap replacement)
+				const writeOrCreateMask = fsnotify.Write | fsnotify.Create
+				if (filepath.Clean(event.Name) == configFile &&
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: Please assign all the conditions into named variabls. e.g. eventFileMatches := filepath.Clean .... == configFile, etc. Then the `if` will be way easier to read. I would read "if X Y Z, set the config".

File: deploy/operator.yaml [hunk 30]
```diff
@@ -50,30 +50,32 @@ spec:
...
             privileged: true
       volumes:
       - name: dbus-socket
         hostPath:
           path: /run/dbus/system_bus_socket
           type: Socket
+      - name: config-volume
+        configMap:
+          # Provide the name of the ConfigMap containing the file
+          # configmap must be created from file it's full path must
```
Comment (phoracek) [inline deploy/operator.yaml]: How is the configmap created is not important. The format is. Could you please provide an example of the body instead?

With something like:

```
data:
  nmstate.yaml: |
    key: value
    key: value

File: pkg/configurations/configurations.go [hunk 31]
```diff
@@ -0,0 +1,159 @@
...
+
+	"github.com/fsnotify/fsnotify"
+
+	"github.com/thoas/go-funk"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+
+	"github.com/spf13/viper"
+)
+
+const defaultInterfacesFilter = "veth*"
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: We can get rid of it and handle it in the initialization of `config`. Just add a comment that we are initializing with default values.

File: pkg/configurations/configurations.go [hunk 32]
```diff
@@ -0,0 +1,159 @@
...
+	v          *viper.Viper
+}
+
+func NewConfigWatcher() *configWatcher {
+	configPathTemp, isSet := os.LookupEnv("CONFIG_PATH")
+	if !isSet {
+		panic("CONFIG_PATH is mandatory")
+	}
+	cw := configWatcher{
+		configPath: configPathTemp,
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: Might as well just set `v` here.

File: pkg/configurations/configurations.go [hunk 33]
```diff
@@ -0,0 +1,159 @@
...
+	}
+	cw.v = viper.New()
+	return &cw
+}
+
+func GetCurrentConfig() config {
+	return c
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: We need the mutex for reading.

File: pkg/configurations/configurations.go [hunk 34]
```diff
@@ -0,0 +1,159 @@
...
+)
+
+// Struct to match config file
+type config struct {
+	NodeNetworkRefreshInterval int    `mapstructure:"node_network_state_refresh_interval"`
+	InterfaceFilter            string `mapstructure:"interfaces_filter"`
+	InterfacesFilterGlob       glob.Glob
+}
+
+var c = config{
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: I'm not a big fan of coupling watcher and config. Responsibilities are not so clear here. What about following:

* Split watcher and configurations to two separate files in this package
* Extend config with getters and setters. Maybe one setter to set the whole thing and one getter per each value. All these will employ a mutex.
* Both watcher and user (readers of loaded values) will use these synchronous/safe getters and setters
* With this, the user facing logic (config alone) will be split from the logic of updating it.

Comment (tsorya) [reply pkg/configurations/configurations.go]: But config setter is loading yaml to struct. How do u see and I need viper for it. It will look really strange but I'll try.

File: pkg/configurations/configurations.go (hunk 33 above)
Comment (tsorya) [reply pkg/configurations/configurations.go]: Why we need mutex for reading? Reading is not atomic?

File: pkg/configurations/configurations.go (hunk 34 above)
Comment (phoracek) [reply pkg/configurations/configurations.go]: Viper has nothing to do with the fact that reload is dynamic, so it can be part of the config struct. Config can have loadFromYaml/File method

File: pkg/configurations/configurations.go (hunk 24 above)
Comment (tsorya) [reply pkg/configurations/configurations.go]: Then I will need 2 go routines

File: pkg/configurations/configurations.go (hunk 33 above)
Comment (phoracek) [reply pkg/configurations/configurations.go]: Not while somebody is writting to it. Afaik go has no guarantees on atomicity of string reading/writting. If you find a trustworthy source saying otherwise, please let me know.

File: pkg/configurations/configurations.go (hunk 24 above)
Comment (phoracek) [reply pkg/configurations/configurations.go]: Oh now I get it. Sorry. Don't mind the comment.

File: pkg/configurations/configurations.go [hunk 35]
```diff
@@ -0,0 +1,159 @@
...
+func (c *configWatcher) watchConfigFile() {
+	configFile := filepath.Clean(c.configPath)
+	configDir, _ := filepath.Split(configFile)
+	realConfigFile, _ := filepath.EvalSymlinks(c.configPath)
+	newWatcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Error(err, "Failed to start fsnotify watcher")
+		return
+	}
+	newWatcher.Add(configDir)
```
Comment (phoracek) [inline pkg/configurations/configurations.go]: Can this fail?

File: deploy/operator.yaml (hunk 30 above)
Comment (tsorya) [reply deploy/operator.yaml]: It is important. If it is not created from file. It will not work.

File: pkg/configurations/configurations.go (hunk 31 above)
Comment (tsorya) [reply pkg/configurations/configurations.go]: but why not to have defaults as consts?

File: pkg/configurations/configurations.go (hunk 34 above)
Comment (tsorya) [reply pkg/configurations/configurations.go]: hmm need to think about it. 
now i am doing it using viper "err := v.Unmarshal(&c)". It is easy and very easily extendable and u ask me to read yaml.
I really don't think there is any need to be so generic and decoupled.

File: pkg/configurations/configurations.go (hunk 31 above)
Comment (phoracek) [reply pkg/configurations/configurations.go]: It's unnecessary duplication. Unless we want to reinitialize the config. Mm, maybe it's just a matter of style. I leave it to you, since I don't really have a convincing explanation.

File: pkg/configurations/configurations.go (hunk 34 above)
Comment (phoracek) [reply pkg/configurations/configurations.go]: I care about the interface for the rest of the application and decoupling of the daemon which reads the current implementation of the config (ConfigMap as a file) from the config object.

File: deploy/operator.yaml (hunk 30 above)
Comment (phoracek) [reply deploy/operator.yaml]: I'm pretty sure user can just create a config map which would be a prepared manifest with populated nmstate.yaml key. I haven't tried it yet, but I believe that --from-file is just a helper in kubectl to do exactly that.

Comment (phoracek) [reply deploy/operator.yaml]: Could you dump ConfigMap with -o yaml for me?

File: pkg/configurations/configurations.go (hunk 35 above)
Comment (tsorya) [reply pkg/configurations/configurations.go]: It can only if fsnotify watcher is closed. In this case we control it so it doesn't need to be handled as for me

File: pkg/configurations/configurations.go (hunk 34 above)
Comment (tsorya) [reply pkg/configurations/configurations.go]: got it

File: deploy/operator.yaml (hunk 30 above)
Comment (tsorya) [reply deploy/operator.yaml]: maybe you are right i will check.

File: pkg/apis/nmstate/v1alpha1/zz_generated.openapi.go [hunk 36]
```diff
@@ -11,15 +11,15 @@ import (
 
 func GetOpenAPIDefinitions(ref common.ReferenceCallback) map[string]common.OpenAPIDefinition {
 	return map[string]common.OpenAPIDefinition{
-		"./pkg/apis/nmstate/v1alpha1.Condition":                               schema_pkg_apis_nmstate_v1alpha1_Condition(ref),
```
Comment (tsorya) [inline pkg/apis/nmstate/v1alpha1/zz_generated.openapi.go]: remove this file

File: deploy/operator.yaml (hunk 30 above)
Comment (tsorya) [reply deploy/operator.yaml]: Checked your right

**Comment by qinqon** [comment]
We have remove nmstate-config since it consumes watchers per handler and at a huge cluster this hits apiserver. Use env vars at handler is good enough, other options will be configured by the operator CRD.
